src/Disks/DiskLocal.cpp

```diff
@@ -409,67 +458,212 @@ void DiskLocal::applyNewSettings(const Poco::Util::AbstractConfiguration & confi
+                disk_checker_magic_number = -1;
+            }
         }
-        else
+    }
+    catch (...)
+    {
+        LOG_ERROR(logger, "We cannot tell if {} exists anymore, or read from it. Most likely disk {} is broken", disk_checker_path, name);
+        throw;
+    }
+
+    /// Try to create a new checker file. The disk status can be either broken or readonly.
+    if (disk_checker_magic_number == -1)
+    try
+    {
+        pcg32_fast rng(randomSeed());
+        UInt32 magic_number = rng();
         {
-            disk->reserved_bytes -= size;
+            auto buf = writeFile(disk_checker_path, DBMS_DEFAULT_BUFFER_SIZE, WriteMode::Rewrite);
+            writeIntBinary(magic_number, *buf);
         }
-
-        if (disk->reservation_count == 0)
-            LOG_ERROR(disk->log, "Unbalanced reservation count for disk '{}'.", disk->getName());
-        else
-            --disk->reservation_count;
+        disk_checker_magic_number = magic_number;
     }
     catch (...)
     {
-        tryLogCurrentException(__PRETTY_FUNCTION__);
+        LOG_WARNING(
+            logger,
+            "Cannot create/write to {0}. Disk {1} is either readonly or broken. Without setting up disk checker file, DiskLocalCheckThread "
+            "will not be started. Disk is assumed to be RW. Try manually fix the disk and do `SYSTEM RESTART DISK {1}`",
+            disk_checker_path,
+            name);
+        disk_checker_can_check_read = false;
+        return true;
     }
-}
 
+    if (disk_checker_magic_number == -1)
+        throw Exception("disk_checker_magic_number is not initialized. It's a bug", ErrorCodes::LOGICAL_ERROR);
+    return true;
+}
 
 void registerDiskLocal(DiskFactory & factory)
 {
     auto creator = [](const String & name,
                       const Poco::Util::AbstractConfiguration & config,
                       const String & config_prefix,
                       ContextPtr context,
-                      const DisksMap & map) -> DiskPtr {
+                      const DisksMap & map) -> DiskPtr
+    {
         String path;
         UInt64 keep_free_space_bytes;
         loadDiskLocalConfig(name, config, config_prefix, context, path, keep_free_space_bytes);
 
         for (const auto & [disk_name, disk_ptr] : map)
-        {
             if (path == disk_ptr->getPath())
-                throw Exception("Disk " + name + " and Disk " + disk_name + " cannot have the same path" + " (" + path + ")", ErrorCodes::BAD_ARGUMENTS);
-        }
-        return std::make_shared<DiskLocal>(name, path, keep_free_space_bytes);
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Disk {} and disk {} cannot have the same path ({})", name, disk_name, path);
+
+        std::shared_ptr<IDisk> disk
+            = std::make_shared<DiskLocal>(name, path, keep_free_space_bytes, context, config.getUInt("local_disk_check_period_ms", 0));
```

Comment:
@amosbird is default value for `local_disk_check_period_ms` sane? This will cause lots of excessive IO, right? (note, that I saw this line in two hunks)

Reply:
Oh, I see, 0 means it is disabled. Sorry for the noise.